community/lucene-index/src/main/java/org/apache/lucene/index/PooledConcurrentMergeScheduler.java

```diff
@@ -0,0 +1,157 @@
+import java.util.concurrent.locks.LockSupport;
+
+import org.neo4j.helpers.NamedThreadFactory;
+import org.neo4j.unsafe.impl.internal.dragons.FeatureToggles;
+
+/**
+ * Lucene indexes merge scheduler that execute merges in a thread pool instead of starting separate thread for each
+ * merge. Each writer have it's own scheduler but all of them use shared pool.
+ *
+ * Current implementation is minimalistic in a number of things it knows about lucene internals to simplify
+ * migrations overhead. It wraps up lucene internal merge tasks and execute them in a common thread pool.
+ * In case if pool and queue exhausted merge will be performed in callers thread.
+ * Since we cant rely on lucene per writer merge threads we need to perform writer tasks counting ourselves to prevent
+ * cases while writer will be closed in the middle of merge and will wait for all writer related merges to complete
+ * before allowing close of writer scheduler.
+ */
+public class PooledConcurrentMergeScheduler extends ConcurrentMergeScheduler
+{
+    private static final int POOL_QUEUE_CAPACITY =
+            FeatureToggles.getInteger( PooledConcurrentMergeScheduler.class, "pool.queue.capacity", 100 );
+    private static final int POOL_MINIMUM_THREADS =
+            FeatureToggles.getInteger( PooledConcurrentMergeScheduler.class, "pool.minimum.threads", 4 );
+
+    private AtomicInteger writerTaskCounter = new AtomicInteger();
+
+    @Override
+    public synchronized void merge( IndexWriter writer, MergeTrigger trigger, boolean newMergesFound )
+            throws IOException
+    {
+        while ( true )
+        {
+            MergePolicy.OneMerge merge = writer.getNextMerge();
+            if ( merge == null )
+            {
+                return;
+            }
+            boolean success = false;
+            try
+            {
+                MergeThread mergeThread = getMergeThread( writer, merge );
+                writerTaskCounter.incrementAndGet();
+                PooledConcurrentMergePool.mergeThreadsPool.submit( mergeTask( mergeThread ) );
+                success = true;
+            }
+            finally
+            {
+                if ( !success )
+                {
+                    writerTaskCounter.decrementAndGet();
+                    writer.mergeFinish( merge );
+                }
+            }
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        waitForAllTasks();
+        super.close();
+    }
+
+    @Override
+    protected void updateMergeThreads()
+    {
+        //noop
+    }
+
+    @Override
+    void removeMergeThread()
+    {
+        // noop
+    }
+
+    AtomicInteger getWriterTaskCounter()
+    {
+        return writerTaskCounter;
+    }
+
+    private Runnable mergeTask( MergeThread mergeThread )
+    {
+        return new MergeTask( mergeThread, writerTaskCounter );
+    }
+
+    private void waitForAllTasks()
+    {
+        while ( writerTaskCounter.get() > 0 )
```

Comment:
Would it make sense to use `Predicates.await()` here and limit max wait time somehow?